lms/djangoapps/discussion_api/views.py

```diff
@@ -568,3 +586,248 @@ def post(self, request):
+        SessionAuthenticationAllowInactiveUser,
+    )
+    parser_classes = (JSONParser, MergePatchParser,)
+    permission_classes = (permissions.IsAuthenticated, permissions.IsAdminUser)
+
+    def _get_representation(self, course, course_key, discussion_settings):
+        """
+        Return a serialized representation of the course discussion settings.
+        """
+        divided_course_wide_discussions, divided_inline_discussions = get_divided_discussions(
+            course, discussion_settings
+        )
+        return JsonResponse({
+            'id': discussion_settings.id,
+            'divided_inline_discussions': divided_inline_discussions,
+            'divided_course_wide_discussions': divided_course_wide_discussions,
+            'always_divide_inline_discussions': discussion_settings.always_divide_inline_discussions,
+            'division_scheme': discussion_settings.division_scheme,
+            'available_division_schemes': available_division_schemes(course_key)
+        })
+
+    def _get_request_kwargs(self, course_id):
+        return dict(course_id=course_id)
+
+    def get(self, request, course_id):
+        """
+        Implement a handler for the GET method.
+        """
+        kwargs = self._get_request_kwargs(course_id)
+        form = CourseDiscussionSettingsForm(kwargs, request_user=request.user)
+
+        if not form.is_valid():
+            raise ValidationError(form.errors)
+
+        course_key = form.cleaned_data['course_key']
+        course = form.cleaned_data['course']
+        discussion_settings = get_course_discussion_settings(course_key)
+        return self._get_representation(course, course_key, discussion_settings)
+
+    def patch(self, request, course_id):
+        """
+        Implement a handler for the PATCH method.
+        """
+        if request.content_type != MergePatchParser.media_type:
+            raise UnsupportedMediaType(request.content_type)
+
+        kwargs = self._get_request_kwargs(course_id)
+        form = CourseDiscussionSettingsForm(kwargs, request_user=request.user)
+        if not form.is_valid():
+            raise ValidationError(form.errors)
+
+        course = form.cleaned_data['course']
+        course_key = form.cleaned_data['course_key']
+        discussion_settings = get_course_discussion_settings(course_key)
+
+        serializer = DiscussionSettingsSerializer(
+            data=request.data,
+            partial=True,
+            course=course,
+            discussion_settings=discussion_settings
+        )
+        if not serializer.is_valid():
+            raise ValidationError(serializer.errors)
+
+        settings_to_change = serializer.validated_data['settings_to_change']
+
+        try:
+            discussion_settings = set_course_discussion_settings(course_key, **settings_to_change)
+        except ValueError as e:
+            raise ValidationError(unicode(e))
+
+        return self._get_representation(course, course_key, discussion_settings)
```

Comment:
The [edX REST doc](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/18350757/edX+REST+API+Conventions#edXRESTAPIConventions-3.HTTPVerbs) mentions a PATCH request must return 204 No content. Do you think it should be better changing this into a PUT or just return 204?

Reply:
@viadanna, this endpoint mimics the behaviour of the existing endpoint, which does the wrong thing. If we just return a 204, an additional GET request is needed to retrieve the updated settings. Is that okay?

Reply:
@viadanna, in terms of semantics, a `PUT` would replace the whole object and not do a partial update. And the code for this method also uses the [JSON Merge Patch](https://tools.ietf.org/html/rfc7396) method and its semantics to do a partial update. [Here](https://openedx.atlassian.net/wiki/spaces/OpenDev/pages/18350757/edX%2BREST%2BAPI%2BConventions#edXRESTAPIConventions-AnoteonHTTPPATCH) is what the new API guidelines state about this. So I think it okay to return a `HTTP 204 No Content` response status code to denote success. If the client needs the updated settings (which can be calculated by the client as well), an additional GET request has to be made. Since the data returned for that request will be small enough (it is just a few fields), I think we can ignore the cost of making an additional request.

Reply:
Is it okay to return 204 and call this done?